Implement Service Checks For Windows

cthiel42 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: ./CONTRIBUTING.md
-->

<!-- _Please make sure to review and check all of these items:_ -->


##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make test-all` (UNIX) passes. CI will also test this
- [ ] unit and/or integration tests are included (if applicable)
- [ ] documentation is changed or added (if applicable)

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Description of change
<!--
Please provide a description of the change here. Be sure to use issue references when applicable:
https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords
-->
I noticed some of the functionality for Windows wasn't implemented when I was using Goss with Windows a few months ago and I'd like to implement some of it, starting with services.

I tried to keep the initial changes here simple to see what you were looking for. For instance, services and processes are set up a little different than the others, so there would need to be some small refactoring if you wanted the windows service file to only be compiled into the windows build. I have an idea about how it could be accomplished somewhat cleanly, but wanted to get your take on whether it was worth it or not first. Same goes for integration tests. There was one test already in there, let me know if you wanted more and what you want them to cover.

Commit:
save off changes

Commit:
save off changes

Commit:
windows file name preventing file from building

Commit:
windows file name preventing file from building

cthiel42 committed:
release builds work when compiled on linux and ran on windows. can start implementing service logic now

cthiel42 committed:
release builds work when compiled on linux and ran on windows. can start implementing service logic now

cthiel42 committed:
testing something

cthiel42 committed:
testing something

cthiel42 committed:
removing

cthiel42 committed:
removing

cthiel42 committed:
windows services working

cthiel42 committed:
windows services working

cthiel42 committed:
remove skip in integration test

cthiel42 committed:
remove skip in integration test

cthiel42 committed:
update exists method

cthiel42 committed:
update exists method

cthiel42 committed:
another test update

cthiel42 committed:
another test update

petemounce commented on system/service_win.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (s *ServiceWindows) Exists() (bool, error) {
+	cmd := util.NewCommand("powershell", "-command", fmt.Sprintf("Get-Service -Name %s", s.service))
```
```suggestion
	cmd := util.NewCommand("powershell", "-NoProfile", "-command", fmt.Sprintf("Get-Service -Name %s", s.service))
```

... so that powershell doesn't spend time loading whose-ever's profile this is, nor is subject to whatever might be inside it.

(it would be good to refactor out a powershell helper method such that all powershell invocations had that flag)

cthiel42 replied:
I made a helper function for calling powershell and including this flag as a default

petemounce commented on system/service_win.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (s *ServiceWindows) Exists() (bool, error) {
+	cmd := util.NewCommand("powershell", "-command", fmt.Sprintf("Get-Service -Name %s", s.service))
```
I forget; is the match greedy? For example, if there are two services `foo1` and `foo2`, will `Get-Service -Name foo` return both?

cthiel42 replied:
It has to be an exact match otherwise it'll return not found.


![pwsh_example](https://github.com/user-attachments/assets/e65cd79b-9e98-4884-9c4e-54aa1c23ca3a)

petemounce commented on system/service_win.go:
```diff
@@ -0,0 +1,54 @@
+func (s *ServiceWindows) Enabled() (bool, error) {
+	cmd := util.NewCommand("powershell", "-command", fmt.Sprintf("$(Get-Service -Name %s).StartType", s.service), "-eq", "\"Automatic\"")
+	cmd.Run()
+	if strings.Contains(cmd.Stdout.String(), "True") {
```
(suggestion only, to make the shell command easier to read by moving the conditional into go)

```suggestion
	cmd := util.NewCommand("powershell", "-command", fmt.Sprintf("$(Get-Service -Name %s).StartType", s.service))
	cmd.Run()
	if strings.Contains(cmd.Stdout.String(), "Automatic") {
```

cthiel42 replied:
This is a lot cleaner so I switched to this.

petemounce commented on system/service_win.go:
```diff
@@ -0,0 +1,54 @@
+func (s *ServiceWindows) Running() (bool, error) {
+	cmd := util.NewCommand("powershell", "-command", fmt.Sprintf("$(Get-Service -Name %s).Status", s.service), "-eq", "\"Running\"")
+	cmd.Run()
+	if strings.Contains(cmd.Stdout.String(), "True") {
```
```suggestion
	cmd := util.NewCommand("powershell", "-command", fmt.Sprintf("$(Get-Service -Name %s).Status", s.service))
	cmd.Run()
	if strings.Contains(cmd.Stdout.String(), "Running") {
```

cthiel42 replied:
I changed and tested this in the latest commit. I like this approach better.

petemounce reviewed (commented):
@cthiel42 thanks very much for your contribution, and apologies for latency on review. This looks broadly good to me. I think ideally it'd be conditionally compiled so it's only present within the Windows binary to follow the pattern the other OS-specific support established.

re: my suggestions - please consider them my own in-memory suggestions for changes as opposed to relying on them to be syntactically / semantically correct. I made them inline during the review, not via an editor.

@aelsabbahy any comments?

cthiel42 commented:
@petemounce Thanks for the review. I'll take a look over the next few days and push some changes for your comments.

cthiel42 committed:
change how powershell commands are called, separate out windows file in builds

cthiel42 committed:
change how powershell commands are called, separate out windows file in builds

cthiel42 commented:
@petemounce I pushed up some changes. I switched the powershell calls to be through a helper function. I removed the comparison operator from the powershell commands and put them in the code (thanks for that suggestion, it looks a lot cleaner to me).  Lastly, the windows service file is now not included in linux builds.

On that last item, I accomplished that by stubbing out the call to NewServiceWindows. That's a functioning way for the linux builds to run with that call in there. Another way would be to take `system.go` and split it into `system_linux.go` and `system_windows.go`. Let me know your thoughts on that.

petemounce commented on system/service_windows.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func (s *ServiceWindows) Enabled() (bool, error) {
+	cmd := util.NewCommandForWindowsPowershell(fmt.Sprintf("$(Get-Service -Name %s).StartType", s.service))
```
```suggestion
	cmd := util.NewCommandForWindowsPowershell(fmt.Sprintf("$(Get-Service -Name %q).StartType", s.service))
```

... to quote the input, making resilient to service names with spaces breaking the invocation and also reduce injection risk (I think...?).

(Again, no local way to try verify that my suggestion is legit; apologies)

cthiel42 replied:
I switched it to %q and it works fine. Service names cannot actually have spaces in them (only the display name can). I think technically there could be an injection risk if quoting is not used, so might as well use it.

petemounce commented on system/service_windows.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func (s *ServiceWindows) Running() (bool, error) {
+	cmd := util.NewCommandForWindowsPowershell(fmt.Sprintf("$(Get-Service -Name %s).Status", s.service))
```
```suggestion
	cmd := util.NewCommandForWindowsPowershell(fmt.Sprintf("$(Get-Service -Name %q).Status", s.service))
```

petemounce commented on util/command_windows.go:
```diff
@@ -28,3 +27,25 @@ func NewCommandForWindowsCmd(name string, arg ...string) *Command {
 }
+
+func NewCommandForWindowsPowershell(name string, arg ...string) *Command {
+	//fmt.Println(arg)
```
```suggestion
```

cthiel42 replied:
this is removed now

petemounce commented on util/command_windows.go:
```diff
@@ -28,3 +27,25 @@ func NewCommandForWindowsCmd(name string, arg ...string) *Command {
+func NewCommandForWindowsPowershell(name string, arg ...string) *Command {
+	//fmt.Println(arg)
+	command := new(Command)
+	command.name = "powershell"
```
Maybe not this PR, but I gather that powershell-core's binary is `pwsh`. 

Is there a way this choice might be later threaded through without a breaking change? I guess it would be a property on the `service` like `powershell-binary`, defaulting to `powershell`...?

cthiel42 replied:
Are you thinking of using pwsh if its available, and falling back to powershell if its not available?

petemounce replied:
No, I was thinking of this being left hard-coded to `powershell` and then (later in a follow-up) providing the ability to a test-author to override that with (something like) `powershell_binary: pwsh` in the test YAML definition. If that test-attribute is omitted, the default no-value choice would be `powershell` as in this PR.

cthiel42 replied:
I haven't looked too much into this, but I think as long as we can get the config to be accessible within this method, we could do that.

petemounce reviewed (commented):
@cthiel42 thanks very much for your increment, and apologies again for latency. Couple more suggestions.

I see I now have an `Approve workflows to run` button, so I've blithely pressed that and can see GHA has done its thing. Not sure if that button is needed each time; will see. Still no Travis, though; I lack access to address that.

I'd be cool with this merging once CI is run and green. @ripienaar / @aelsabbahy, any thoughts?

cthiel42 committed:
quoted strings and remove unneeded comment

cthiel42 committed:
quoted strings and remove unneeded comment

petemounce commented:
@cthiel42 would you mind addressing https://github.com/goss-org/goss/actions/runs/21811022610/job/63155554726?pr=1017#step:4:34, which ought to just be a `go fmt` to make it progress?

cthiel42 committed:
gofmt

cthiel42 committed:
gofmt

cthiel42 commented:
> @cthiel42 would you mind addressing https://github.com/goss-org/goss/actions/runs/21811022610/job/63155554726?pr=1017#step:4:34, which ought to just be a `go fmt` to make it progress?

@petemounce This should hopefully be resolved now
